app/controllers/application_controller.rb

```diff
@@ -4,9 +4,10 @@ class ApplicationController < ActionController::Base
   rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
 
   private
-    def user_not_authorized
-      @error = "User not authorized to perform this action"
-      flash[:error] = @error
-      redirect_to controller: "products", action: "new"
+    def user_not_authorized(exception)
+      policy_name = exception.policy.class.to_s.underscore
+      dynamic_controller = policy_name.split("_")[0]
+      flash[:error] = "#{policy_name}.#{exception.query} not allowed for your role"
+      redirect_to controller: 'home', action: 'index'
```

Comment:
You can replace it something to go back to previous page

redirect_back(fallback_location: root_path)